### DISCUSSION TITLE
[General] Proof Generation and Submission for Mainnet

### DISCUSSION THREAD

**Discussion by sagarparker** [zkSync-Community-Hub/zksync-developers#858] [General] [answered]
### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement and the issue you are running into.

Hi,

I have been working on running the prover subsystem and successfully replicated the proving batch process as outlined in the documentation: https://matter-labs.github.io/zksync-era/prover/latest/05_proving_batch.html

I wanted to inquire if there is a way to generate and submit proofs to the gateway for mainnet without using the ExternalProofIntegrationAPI. Specifically, is it possible to fetch proof generation data directly for mainnet from the gateway?

Looking forward to your guidance.

**Comment by alhenaai** [comment]
Hi,

To generate and submit proofs to the mainnet gateway without using the ExternalProofIntegrationAPI, you can directly fetch proof generation data using the `/proof_generation_data` endpoint. This endpoint allows you to retrieve the latest batch proof inputs, which are very similar to how ZKsync's provers work internally. You can also fetch inputs for a specific batch proof using `/proof_generation_data/{batch_number}` for testing or benchmarking purposes.

For more details on how to use these endpoints and manage proof submissions, you can refer to the documentation on the Prover API for ZKsync [here](https://zksync.mirror.xyz/1HzbsDVMQeE3P-DgysjKzagNnHehwCIrC1QA9XcNRnk).

For further discussions and support, you might find it helpful to visit the ZKsync developers discussions on Github at [ZKsync Developers Discussions](https://github.com/zkSync-Community-Hub/zksync-developers/discussions).

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Reply by sagarparker** [reply]
Hi,

I have already followed the process outlined in the documentation and successfully used the following API:

```
wget --content-disposition https://prover-api.zksync.io/proof_generation_data  
```

However, I was wondering if there is a way to perform this process without relying on the witness input from the API. Specifically, is it possible to obtain the required data for mainnet directly from the gateway?

Looking forward to your insights.

**Reply by alhenaai** [reply]
Hi,

Currently, the process for obtaining the necessary data for proof generation in ZKsync, especially for mainnet, relies on using the Prover API to fetch witness inputs. The endpoint you mentioned, `https://prover-api.zksync.io/proof_generation_data`, is the standard method to retrieve this data.

If you are looking to bypass the API and obtain data directly from the gateway, there isn't a documented method available that supports this approach. The Prover API is designed to ensure that the data you receive is accurate and formatted correctly for generating proofs.

For any updates or alternative methods that might be introduced, I recommend keeping an eye on the official ZKsync documentation or the ZKsync developers discussions on Github at [ZKsync Developers Discussions](https://github.com/zkSync-Community-Hub/zksync-developers/discussions) for the latest information and community insights.

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Reply by sagarparker** [reply]
Hi,

Could a developer or someone from the team kindly confirm this?

Thank you for your assistance.

**Reply by alhenaai** [reply]
For confirmation and further insights directly from the ZKsync development team or other developers, I recommend posting your query in the ZKsync developers discussions on Github. This platform is specifically designed for developers to discuss, share insights, and get direct responses from the ZKsync team and community.

You can access the discussions and post your question [here](https://github.com/zkSync-Community-Hub/zksync-developers/discussions). This will allow you to engage directly with the team and other developers who might have explored similar queries or have insights into your specific needs.

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Comment by EmilLuta** [comment]
Hey @sagarparker, this is Emil -- I work on Provers.

It's unclear to me what you mean by "gateway". Is it [Prover Gateway](https://github.com/matter-labs/zksync-era/tree/main/prover/crates/bin/prover_fri_gateway) that you refer to? If yes, we don't support getting the information from there (albeit, it would be 100% identical with what you see on External Prover API). If not, please advise on gateway and I'll come back with a more accurate answer.

Cheers!

**Reply by sagarparker** [reply]
Hi Emil,

Yes, I was referring to the Prover Gateway. Is the external Prover API the only method for obtaining the witness input data for Mainnet currently?

**Reply by EmilLuta** [reply, marked as answer]
Yes, that's correct. And note that prover gateway gets exactly the same information that you would get from External Prover API. There's no difference on data retrieval (prover gateway will request information from an Internal Prover API, whilst external actors can use the External Prover API -- both of which share the same code). Is there any reason you'd be interested in Prover Gateway's data?

**Reply by sagarparker** [reply]
Understood, thanks! I was trying to run the proof generation process without the need of using external prover API.
